# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [3]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    source_id_text = []
    target_id_text = []

    
    for sentence in source_text.split("\n"):
        out = [source_vocab_to_int[word] for word in sentence.split()]
        source_id_text.append(out)
    for sentence in target_text.split("\n"):
        out = [target_vocab_to_int[word] for word in sentence.split()]
        out.append(target_vocab_to_int['<EOS>'])
        target_id_text.append(out)
            
    return source_id_text, target_id_text
            


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper
import problem_unittests as tests

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.1.0


/home/erimel/anaconda3/envs/translate/lib/python3.6/site-packages/ipykernel_launcher.py:15: UserWarning: No GPU found. Please use a GPU to train your neural network.
  from ipykernel import kernelapp as app


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [7]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    # TODO: Implement Function
    
    input = tf.placeholder(tf.int32, shape=[None, None], name="input")
    targets = tf.placeholder(tf.int32, shape=[None, None], name="targets")
    learning_rate = tf.placeholder(tf.float32, name= "learning_rate")
    keep_prob = tf.placeholder(tf.float32, name= "keep_prob")
    target_sequence_length = tf.placeholder(tf.int32, shape=[None], name= "target_sequence_length")
    max_target_len = tf.reduce_max(target_sequence_length, name="max_target_len")
    source_sequence_length = tf.placeholder(tf.int32, shape=[None], name= "source_sequence_length")
    return input, targets, learning_rate, keep_prob, target_sequence_length, max_target_len, source_sequence_length


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [8]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # TODO: Implement Function
    go_id = tf.constant(target_vocab_to_int['<GO>'], shape=(batch_size,1), dtype=tf.int32)
    # Concatenate the vector without the last word id with the go ids vector
    processed_input = tf.concat([go_id,target_data[:,:-1]],1)
    return processed_input

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [33]:
from imp import reload
reload(tests)

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    # TODO: Implement Function
    def build_lstm_cell(rnn_size, keep_prob):
        lstm = tf.contrib.rnn.LSTMCell(rnn_size)
        #lstm_drop = tf.contrib.rnn.DropoutWrapper(lstm, output_keep_prob=keep_prob)
        return lstm
    stacked_lstm = tf.contrib.rnn.MultiRNNCell([build_lstm_cell(rnn_size, keep_prob) for _ in range(num_layers)])
    lstm_drop = tf.contrib.rnn.DropoutWrapper(stacked_lstm, output_keep_prob=keep_prob)
    embed_encoder = tf.contrib.layers.embed_sequence(rnn_inputs, vocab_size = source_vocab_size, embed_dim = encoding_embedding_size)
    output, state = tf.nn.dynamic_rnn(stacked_lstm, embed_encoder, source_sequence_length, dtype=tf.float32)
    return output, state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [38]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    # TODO: Implement Function
    trainig_helper = tf.contrib.seq2seq.TrainingHelper(dec_embed_input, target_sequence_length)
    basic_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell, trainig_helper, encoder_state, output_layer)
    final_outputs, final_state = tf.contrib.seq2seq.dynamic_decode(basic_decoder, maximum_iterations=max_summary_length)
    return final_outputs
    

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [39]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    # TODO: Implement Function
    
    start_ids = tf.tile([start_of_sequence_id], [batch_size])
    embedding_helper = tf.contrib.seq2seq.GreedyEmbeddingHelper(dec_embeddings, start_ids, end_of_sequence_id)
    basic_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell, embedding_helper, encoder_state, output_layer)
    final_outputs, final_state = tf.contrib.seq2seq.dynamic_decode(basic_decoder, maximum_iterations=max_target_sequence_length)
    return final_outputs




"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [40]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :param decoding_embedding_size: Decoding embedding size
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    def build_cell(rnn_size, keep_prob):
        lstm = tf.contrib.rnn.LSTMCell(rnn_size)
        lstm_drop = tf.contrib.rnn.DropoutWrapper(lstm, output_keep_prob=keep_prob)
        return lstm_drop
    # Stack them all
    stacked_lstm = tf.contrib.rnn.MultiRNNCell([build_cell(rnn_size, keep_prob) for _ in range(num_layers)])
    
    dec_embeddings = tf.Variable(tf.random_uniform([target_vocab_size, decoding_embedding_size]))
    dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, dec_input)

    dense_layer = Dense(target_vocab_size,
                         kernel_initializer = tf.truncated_normal_initializer(mean = 0.0, stddev=0.1))
    
    with tf.variable_scope("decode") as scope:
        tr_decoder_output = decoding_layer_train(
            encoder_state, stacked_lstm, dec_embed_input, 
            target_sequence_length, max_target_sequence_length, 
            dense_layer, keep_prob)
        scope.reuse_variables()
        inf_decoder_output = decoding_layer_infer(
            encoder_state, stacked_lstm, dec_embeddings, 
            target_vocab_to_int['<GO>'], target_vocab_to_int['<EOS>'], 
            max_target_sequence_length, target_vocab_size, 
            dense_layer, batch_size, keep_prob)
    
    return tr_decoder_output, inf_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [42]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    output, state = encoding_layer(input_data, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   enc_embedding_size)
    
    processed_input = process_decoder_input(target_data, target_vocab_to_int, batch_size)
    
    tr_decoder_output, inf_decoder_output = decoding_layer(processed_input, state,
                   target_sequence_length, max_target_sentence_length,
                   rnn_size, num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, dec_embedding_size)
    
    return tr_decoder_output, inf_decoder_output


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [43]:
# Number of Epochs
epochs = 6
# Batch Size
batch_size = 256
# RNN Size
rnn_size = 128
# Number of Layers
num_layers = 2
# Embedding Size
encoding_embedding_size = 30
decoding_embedding_size = 30
# Learning Rate
learning_rate = 0.005
# Dropout Keep Probability
keep_probability = 0.5
display_step = True

### Build the Graph
Build the graph using the neural network you implemented.

In [44]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [45]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [46]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch    1/538 - Train Accuracy: 0.2316, Validation Accuracy: 0.3157, Loss: 5.5012
Epoch   0 Batch    2/538 - Train Accuracy: 0.2639, Validation Accuracy: 0.3276, Loss: 4.7801
Epoch   0 Batch    3/538 - Train Accuracy: 0.2631, Validation Accuracy: 0.3455, Loss: 4.5009
Epoch   0 Batch    4/538 - Train Accuracy: 0.2758, Validation Accuracy: 0.3471, Loss: 4.1871
Epoch   0 Batch    5/538 - Train Accuracy: 0.2984, Validation Accuracy: 0.3473, Loss: 3.8761
Epoch   0 Batch    6/538 - Train Accuracy: 0.3344, Validation Accuracy: 0.3770, Loss: 3.7381
Epoch   0 Batch    7/538 - Train Accuracy: 0.3084, Validation Accuracy: 0.3780, Loss: 3.7138
Epoch   0 Batch    8/538 - Train Accuracy: 0.3109, Validation Accuracy: 0.3770, Loss: 3.6688
Epoch   0 Batch    9/538 - Train Accuracy: 0.3111, Validation Accuracy: 0.3750, Loss: 3.5522
Epoch   0 Batch   10/538 - Train Accuracy: 0.2826, Validation Accuracy: 0.3633, Loss: 3.6027
Epoch   0 Batch   11/538 - Train Accuracy: 0.2932, Validation Accuracy

Epoch   0 Batch   90/538 - Train Accuracy: 0.4498, Validation Accuracy: 0.4750, Loss: 1.7556
Epoch   0 Batch   91/538 - Train Accuracy: 0.4357, Validation Accuracy: 0.4858, Loss: 1.8311
Epoch   0 Batch   92/538 - Train Accuracy: 0.4424, Validation Accuracy: 0.4851, Loss: 1.7521
Epoch   0 Batch   93/538 - Train Accuracy: 0.4193, Validation Accuracy: 0.4856, Loss: 1.7841
Epoch   0 Batch   94/538 - Train Accuracy: 0.4182, Validation Accuracy: 0.4773, Loss: 1.8038
Epoch   0 Batch   95/538 - Train Accuracy: 0.4695, Validation Accuracy: 0.4806, Loss: 1.6229
Epoch   0 Batch   96/538 - Train Accuracy: 0.4686, Validation Accuracy: 0.4805, Loss: 1.6215
Epoch   0 Batch   97/538 - Train Accuracy: 0.4207, Validation Accuracy: 0.4771, Loss: 1.7178
Epoch   0 Batch   98/538 - Train Accuracy: 0.4522, Validation Accuracy: 0.4714, Loss: 1.5798
Epoch   0 Batch   99/538 - Train Accuracy: 0.4141, Validation Accuracy: 0.4711, Loss: 1.7085
Epoch   0 Batch  100/538 - Train Accuracy: 0.4387, Validation Accuracy

Epoch   0 Batch  179/538 - Train Accuracy: 0.4949, Validation Accuracy: 0.5431, Loss: 1.0756
Epoch   0 Batch  180/538 - Train Accuracy: 0.5246, Validation Accuracy: 0.5408, Loss: 1.0293
Epoch   0 Batch  181/538 - Train Accuracy: 0.4697, Validation Accuracy: 0.5327, Loss: 1.1059
Epoch   0 Batch  182/538 - Train Accuracy: 0.4670, Validation Accuracy: 0.5302, Loss: 1.0707
Epoch   0 Batch  183/538 - Train Accuracy: 0.4855, Validation Accuracy: 0.5240, Loss: 0.9986
Epoch   0 Batch  184/538 - Train Accuracy: 0.4777, Validation Accuracy: 0.5279, Loss: 1.0165
Epoch   0 Batch  185/538 - Train Accuracy: 0.5023, Validation Accuracy: 0.5419, Loss: 1.0241
Epoch   0 Batch  186/538 - Train Accuracy: 0.4898, Validation Accuracy: 0.5304, Loss: 1.0213
Epoch   0 Batch  187/538 - Train Accuracy: 0.5236, Validation Accuracy: 0.5284, Loss: 0.9881
Epoch   0 Batch  188/538 - Train Accuracy: 0.4637, Validation Accuracy: 0.5185, Loss: 1.0343
Epoch   0 Batch  189/538 - Train Accuracy: 0.4861, Validation Accuracy

Epoch   0 Batch  268/538 - Train Accuracy: 0.5132, Validation Accuracy: 0.5428, Loss: 0.7739
Epoch   0 Batch  269/538 - Train Accuracy: 0.5143, Validation Accuracy: 0.5597, Loss: 0.8211
Epoch   0 Batch  270/538 - Train Accuracy: 0.5338, Validation Accuracy: 0.5543, Loss: 0.8416
Epoch   0 Batch  271/538 - Train Accuracy: 0.5133, Validation Accuracy: 0.5410, Loss: 0.8373
Epoch   0 Batch  272/538 - Train Accuracy: 0.5004, Validation Accuracy: 0.5394, Loss: 0.8890
Epoch   0 Batch  273/538 - Train Accuracy: 0.5506, Validation Accuracy: 0.5593, Loss: 0.8340
Epoch   0 Batch  274/538 - Train Accuracy: 0.5002, Validation Accuracy: 0.5460, Loss: 0.8763
Epoch   0 Batch  275/538 - Train Accuracy: 0.5125, Validation Accuracy: 0.5385, Loss: 0.8404
Epoch   0 Batch  276/538 - Train Accuracy: 0.5441, Validation Accuracy: 0.5591, Loss: 0.8260
Epoch   0 Batch  277/538 - Train Accuracy: 0.5525, Validation Accuracy: 0.5675, Loss: 0.8098
Epoch   0 Batch  278/538 - Train Accuracy: 0.5541, Validation Accuracy

Epoch   0 Batch  357/538 - Train Accuracy: 0.5844, Validation Accuracy: 0.5913, Loss: 0.7168
Epoch   0 Batch  358/538 - Train Accuracy: 0.5680, Validation Accuracy: 0.5916, Loss: 0.7302
Epoch   0 Batch  359/538 - Train Accuracy: 0.5939, Validation Accuracy: 0.5893, Loss: 0.7068
Epoch   0 Batch  360/538 - Train Accuracy: 0.5941, Validation Accuracy: 0.5923, Loss: 0.7282
Epoch   0 Batch  361/538 - Train Accuracy: 0.5831, Validation Accuracy: 0.5890, Loss: 0.6938
Epoch   0 Batch  362/538 - Train Accuracy: 0.6096, Validation Accuracy: 0.6058, Loss: 0.6816
Epoch   0 Batch  363/538 - Train Accuracy: 0.5984, Validation Accuracy: 0.6012, Loss: 0.6717
Epoch   0 Batch  364/538 - Train Accuracy: 0.5578, Validation Accuracy: 0.6035, Loss: 0.7401
Epoch   0 Batch  365/538 - Train Accuracy: 0.5727, Validation Accuracy: 0.6025, Loss: 0.7033
Epoch   0 Batch  366/538 - Train Accuracy: 0.5945, Validation Accuracy: 0.5969, Loss: 0.7262
Epoch   0 Batch  367/538 - Train Accuracy: 0.5926, Validation Accuracy

Epoch   0 Batch  446/538 - Train Accuracy: 0.6384, Validation Accuracy: 0.6401, Loss: 0.6111
Epoch   0 Batch  447/538 - Train Accuracy: 0.5969, Validation Accuracy: 0.6424, Loss: 0.6530
Epoch   0 Batch  448/538 - Train Accuracy: 0.6365, Validation Accuracy: 0.6413, Loss: 0.6126
Epoch   0 Batch  449/538 - Train Accuracy: 0.6080, Validation Accuracy: 0.6341, Loss: 0.6628
Epoch   0 Batch  450/538 - Train Accuracy: 0.6287, Validation Accuracy: 0.6241, Loss: 0.6420
Epoch   0 Batch  451/538 - Train Accuracy: 0.6182, Validation Accuracy: 0.6204, Loss: 0.6363
Epoch   0 Batch  452/538 - Train Accuracy: 0.6174, Validation Accuracy: 0.6309, Loss: 0.6179
Epoch   0 Batch  453/538 - Train Accuracy: 0.5895, Validation Accuracy: 0.6310, Loss: 0.6531
Epoch   0 Batch  454/538 - Train Accuracy: 0.6291, Validation Accuracy: 0.6373, Loss: 0.6175
Epoch   0 Batch  455/538 - Train Accuracy: 0.6552, Validation Accuracy: 0.6381, Loss: 0.5855
Epoch   0 Batch  456/538 - Train Accuracy: 0.6873, Validation Accuracy

Epoch   0 Batch  535/538 - Train Accuracy: 0.6583, Validation Accuracy: 0.6584, Loss: 0.5790
Epoch   0 Batch  536/538 - Train Accuracy: 0.6492, Validation Accuracy: 0.6564, Loss: 0.5947
Epoch   1 Batch    1/538 - Train Accuracy: 0.6477, Validation Accuracy: 0.6577, Loss: 0.5996
Epoch   1 Batch    2/538 - Train Accuracy: 0.6354, Validation Accuracy: 0.6591, Loss: 0.6150
Epoch   1 Batch    3/538 - Train Accuracy: 0.6094, Validation Accuracy: 0.6559, Loss: 0.5987
Epoch   1 Batch    4/538 - Train Accuracy: 0.6615, Validation Accuracy: 0.6603, Loss: 0.5823
Epoch   1 Batch    5/538 - Train Accuracy: 0.6267, Validation Accuracy: 0.6600, Loss: 0.6024
Epoch   1 Batch    6/538 - Train Accuracy: 0.6592, Validation Accuracy: 0.6603, Loss: 0.5683
Epoch   1 Batch    7/538 - Train Accuracy: 0.6455, Validation Accuracy: 0.6580, Loss: 0.5671
Epoch   1 Batch    8/538 - Train Accuracy: 0.6543, Validation Accuracy: 0.6637, Loss: 0.5927
Epoch   1 Batch    9/538 - Train Accuracy: 0.6521, Validation Accuracy

Epoch   1 Batch   88/538 - Train Accuracy: 0.6752, Validation Accuracy: 0.6880, Loss: 0.5368
Epoch   1 Batch   89/538 - Train Accuracy: 0.6713, Validation Accuracy: 0.6884, Loss: 0.5369
Epoch   1 Batch   90/538 - Train Accuracy: 0.6650, Validation Accuracy: 0.6820, Loss: 0.5358
Epoch   1 Batch   91/538 - Train Accuracy: 0.6709, Validation Accuracy: 0.6873, Loss: 0.5489
Epoch   1 Batch   92/538 - Train Accuracy: 0.6561, Validation Accuracy: 0.6857, Loss: 0.5347
Epoch   1 Batch   93/538 - Train Accuracy: 0.6660, Validation Accuracy: 0.6863, Loss: 0.5427
Epoch   1 Batch   94/538 - Train Accuracy: 0.6762, Validation Accuracy: 0.6847, Loss: 0.5275
Epoch   1 Batch   95/538 - Train Accuracy: 0.6921, Validation Accuracy: 0.6772, Loss: 0.4844
Epoch   1 Batch   96/538 - Train Accuracy: 0.6819, Validation Accuracy: 0.6781, Loss: 0.5044
Epoch   1 Batch   97/538 - Train Accuracy: 0.6617, Validation Accuracy: 0.6834, Loss: 0.5290
Epoch   1 Batch   98/538 - Train Accuracy: 0.6907, Validation Accuracy

Epoch   1 Batch  177/538 - Train Accuracy: 0.6962, Validation Accuracy: 0.7015, Loss: 0.4708
Epoch   1 Batch  178/538 - Train Accuracy: 0.7085, Validation Accuracy: 0.7054, Loss: 0.4600
Epoch   1 Batch  179/538 - Train Accuracy: 0.6893, Validation Accuracy: 0.7033, Loss: 0.4851
Epoch   1 Batch  180/538 - Train Accuracy: 0.7102, Validation Accuracy: 0.7045, Loss: 0.4752
Epoch   1 Batch  181/538 - Train Accuracy: 0.6813, Validation Accuracy: 0.7053, Loss: 0.5001
Epoch   1 Batch  182/538 - Train Accuracy: 0.6545, Validation Accuracy: 0.7010, Loss: 0.4921
Epoch   1 Batch  183/538 - Train Accuracy: 0.7199, Validation Accuracy: 0.7001, Loss: 0.4395
Epoch   1 Batch  184/538 - Train Accuracy: 0.6918, Validation Accuracy: 0.7024, Loss: 0.4522
Epoch   1 Batch  185/538 - Train Accuracy: 0.6979, Validation Accuracy: 0.7051, Loss: 0.4655
Epoch   1 Batch  186/538 - Train Accuracy: 0.6957, Validation Accuracy: 0.7049, Loss: 0.4690
Epoch   1 Batch  187/538 - Train Accuracy: 0.6996, Validation Accuracy

Epoch   1 Batch  266/538 - Train Accuracy: 0.7139, Validation Accuracy: 0.7244, Loss: 0.4323
Epoch   1 Batch  267/538 - Train Accuracy: 0.7041, Validation Accuracy: 0.7250, Loss: 0.4145
Epoch   1 Batch  268/538 - Train Accuracy: 0.7349, Validation Accuracy: 0.7251, Loss: 0.4044
Epoch   1 Batch  269/538 - Train Accuracy: 0.6977, Validation Accuracy: 0.7257, Loss: 0.4224
Epoch   1 Batch  270/538 - Train Accuracy: 0.7131, Validation Accuracy: 0.7278, Loss: 0.4255
Epoch   1 Batch  271/538 - Train Accuracy: 0.7145, Validation Accuracy: 0.7244, Loss: 0.4302
Epoch   1 Batch  272/538 - Train Accuracy: 0.6785, Validation Accuracy: 0.7232, Loss: 0.4569
Epoch   1 Batch  273/538 - Train Accuracy: 0.7041, Validation Accuracy: 0.7290, Loss: 0.4156
Epoch   1 Batch  274/538 - Train Accuracy: 0.6785, Validation Accuracy: 0.7276, Loss: 0.4506
Epoch   1 Batch  275/538 - Train Accuracy: 0.6986, Validation Accuracy: 0.7285, Loss: 0.4426
Epoch   1 Batch  276/538 - Train Accuracy: 0.7271, Validation Accuracy

Epoch   1 Batch  355/538 - Train Accuracy: 0.7293, Validation Accuracy: 0.7440, Loss: 0.3944
Epoch   1 Batch  356/538 - Train Accuracy: 0.7379, Validation Accuracy: 0.7470, Loss: 0.3513
Epoch   1 Batch  357/538 - Train Accuracy: 0.7414, Validation Accuracy: 0.7507, Loss: 0.3777
Epoch   1 Batch  358/538 - Train Accuracy: 0.7285, Validation Accuracy: 0.7512, Loss: 0.3892
Epoch   1 Batch  359/538 - Train Accuracy: 0.7351, Validation Accuracy: 0.7512, Loss: 0.3693
Epoch   1 Batch  360/538 - Train Accuracy: 0.7201, Validation Accuracy: 0.7482, Loss: 0.3904
Epoch   1 Batch  361/538 - Train Accuracy: 0.7476, Validation Accuracy: 0.7511, Loss: 0.3769
Epoch   1 Batch  362/538 - Train Accuracy: 0.7558, Validation Accuracy: 0.7505, Loss: 0.3548
Epoch   1 Batch  363/538 - Train Accuracy: 0.7353, Validation Accuracy: 0.7468, Loss: 0.3522
Epoch   1 Batch  364/538 - Train Accuracy: 0.6977, Validation Accuracy: 0.7452, Loss: 0.4037
Epoch   1 Batch  365/538 - Train Accuracy: 0.7282, Validation Accuracy

Epoch   1 Batch  444/538 - Train Accuracy: 0.7887, Validation Accuracy: 0.7757, Loss: 0.3316
Epoch   1 Batch  445/538 - Train Accuracy: 0.7541, Validation Accuracy: 0.7725, Loss: 0.3380
Epoch   1 Batch  446/538 - Train Accuracy: 0.7662, Validation Accuracy: 0.7720, Loss: 0.3289
Epoch   1 Batch  447/538 - Train Accuracy: 0.7170, Validation Accuracy: 0.7749, Loss: 0.3654
Epoch   1 Batch  448/538 - Train Accuracy: 0.7587, Validation Accuracy: 0.7759, Loss: 0.3184
Epoch   1 Batch  449/538 - Train Accuracy: 0.7643, Validation Accuracy: 0.7736, Loss: 0.3576
Epoch   1 Batch  450/538 - Train Accuracy: 0.7574, Validation Accuracy: 0.7720, Loss: 0.3544
Epoch   1 Batch  451/538 - Train Accuracy: 0.7379, Validation Accuracy: 0.7688, Loss: 0.3395
Epoch   1 Batch  452/538 - Train Accuracy: 0.7557, Validation Accuracy: 0.7706, Loss: 0.3432
Epoch   1 Batch  453/538 - Train Accuracy: 0.7297, Validation Accuracy: 0.7633, Loss: 0.3502
Epoch   1 Batch  454/538 - Train Accuracy: 0.7435, Validation Accuracy

Epoch   1 Batch  533/538 - Train Accuracy: 0.7607, Validation Accuracy: 0.7832, Loss: 0.3245
Epoch   1 Batch  534/538 - Train Accuracy: 0.7474, Validation Accuracy: 0.7818, Loss: 0.3044
Epoch   1 Batch  535/538 - Train Accuracy: 0.7714, Validation Accuracy: 0.7763, Loss: 0.3269
Epoch   1 Batch  536/538 - Train Accuracy: 0.7563, Validation Accuracy: 0.7759, Loss: 0.3219
Epoch   2 Batch    1/538 - Train Accuracy: 0.7650, Validation Accuracy: 0.7747, Loss: 0.3241
Epoch   2 Batch    2/538 - Train Accuracy: 0.7553, Validation Accuracy: 0.7741, Loss: 0.3361
Epoch   2 Batch    3/538 - Train Accuracy: 0.7275, Validation Accuracy: 0.7704, Loss: 0.3209
Epoch   2 Batch    4/538 - Train Accuracy: 0.7572, Validation Accuracy: 0.7681, Loss: 0.3316
Epoch   2 Batch    5/538 - Train Accuracy: 0.7292, Validation Accuracy: 0.7711, Loss: 0.3251
Epoch   2 Batch    6/538 - Train Accuracy: 0.7608, Validation Accuracy: 0.7733, Loss: 0.3107
Epoch   2 Batch    7/538 - Train Accuracy: 0.7471, Validation Accuracy

Epoch   2 Batch   86/538 - Train Accuracy: 0.7625, Validation Accuracy: 0.7598, Loss: 0.2849
Epoch   2 Batch   87/538 - Train Accuracy: 0.7832, Validation Accuracy: 0.7651, Loss: 0.2781
Epoch   2 Batch   88/538 - Train Accuracy: 0.8000, Validation Accuracy: 0.7612, Loss: 0.2869
Epoch   2 Batch   89/538 - Train Accuracy: 0.7688, Validation Accuracy: 0.7688, Loss: 0.2796
Epoch   2 Batch   90/538 - Train Accuracy: 0.7747, Validation Accuracy: 0.7635, Loss: 0.2811
Epoch   2 Batch   91/538 - Train Accuracy: 0.7715, Validation Accuracy: 0.7670, Loss: 0.2840
Epoch   2 Batch   92/538 - Train Accuracy: 0.7605, Validation Accuracy: 0.7626, Loss: 0.2791
Epoch   2 Batch   93/538 - Train Accuracy: 0.7611, Validation Accuracy: 0.7692, Loss: 0.2822
Epoch   2 Batch   94/538 - Train Accuracy: 0.7852, Validation Accuracy: 0.7663, Loss: 0.2766
Epoch   2 Batch   95/538 - Train Accuracy: 0.7832, Validation Accuracy: 0.7638, Loss: 0.2533
Epoch   2 Batch   96/538 - Train Accuracy: 0.7846, Validation Accuracy

Epoch   2 Batch  175/538 - Train Accuracy: 0.7939, Validation Accuracy: 0.8034, Loss: 0.2542
Epoch   2 Batch  176/538 - Train Accuracy: 0.7902, Validation Accuracy: 0.8065, Loss: 0.2679
Epoch   2 Batch  177/538 - Train Accuracy: 0.7870, Validation Accuracy: 0.7949, Loss: 0.2450
Epoch   2 Batch  178/538 - Train Accuracy: 0.7945, Validation Accuracy: 0.8052, Loss: 0.2428
Epoch   2 Batch  179/538 - Train Accuracy: 0.7998, Validation Accuracy: 0.8137, Loss: 0.2524
Epoch   2 Batch  180/538 - Train Accuracy: 0.8212, Validation Accuracy: 0.8022, Loss: 0.2409
Epoch   2 Batch  181/538 - Train Accuracy: 0.7998, Validation Accuracy: 0.8059, Loss: 0.2623
Epoch   2 Batch  182/538 - Train Accuracy: 0.7750, Validation Accuracy: 0.8031, Loss: 0.2393
Epoch   2 Batch  183/538 - Train Accuracy: 0.8283, Validation Accuracy: 0.7882, Loss: 0.2255
Epoch   2 Batch  184/538 - Train Accuracy: 0.8051, Validation Accuracy: 0.8137, Loss: 0.2361
Epoch   2 Batch  185/538 - Train Accuracy: 0.8127, Validation Accuracy

Epoch   2 Batch  264/538 - Train Accuracy: 0.7998, Validation Accuracy: 0.8164, Loss: 0.2220
Epoch   2 Batch  265/538 - Train Accuracy: 0.8096, Validation Accuracy: 0.8176, Loss: 0.2323
Epoch   2 Batch  266/538 - Train Accuracy: 0.8151, Validation Accuracy: 0.8203, Loss: 0.2238
Epoch   2 Batch  267/538 - Train Accuracy: 0.8148, Validation Accuracy: 0.8088, Loss: 0.2163
Epoch   2 Batch  268/538 - Train Accuracy: 0.8266, Validation Accuracy: 0.8159, Loss: 0.2034
Epoch   2 Batch  269/538 - Train Accuracy: 0.8221, Validation Accuracy: 0.8152, Loss: 0.2162
Epoch   2 Batch  270/538 - Train Accuracy: 0.8346, Validation Accuracy: 0.8113, Loss: 0.2201
Epoch   2 Batch  271/538 - Train Accuracy: 0.8416, Validation Accuracy: 0.8137, Loss: 0.2235
Epoch   2 Batch  272/538 - Train Accuracy: 0.7941, Validation Accuracy: 0.8123, Loss: 0.2372
Epoch   2 Batch  273/538 - Train Accuracy: 0.8227, Validation Accuracy: 0.8034, Loss: 0.2201
Epoch   2 Batch  274/538 - Train Accuracy: 0.7871, Validation Accuracy

Epoch   2 Batch  353/538 - Train Accuracy: 0.8309, Validation Accuracy: 0.8297, Loss: 0.2037
Epoch   2 Batch  354/538 - Train Accuracy: 0.8266, Validation Accuracy: 0.8329, Loss: 0.2050
Epoch   2 Batch  355/538 - Train Accuracy: 0.8250, Validation Accuracy: 0.8205, Loss: 0.2025
Epoch   2 Batch  356/538 - Train Accuracy: 0.8562, Validation Accuracy: 0.8278, Loss: 0.1811
Epoch   2 Batch  357/538 - Train Accuracy: 0.8588, Validation Accuracy: 0.8379, Loss: 0.1870
Epoch   2 Batch  358/538 - Train Accuracy: 0.8443, Validation Accuracy: 0.8420, Loss: 0.1947
Epoch   2 Batch  359/538 - Train Accuracy: 0.8438, Validation Accuracy: 0.8462, Loss: 0.1925
Epoch   2 Batch  360/538 - Train Accuracy: 0.8369, Validation Accuracy: 0.8363, Loss: 0.1997
Epoch   2 Batch  361/538 - Train Accuracy: 0.8484, Validation Accuracy: 0.8386, Loss: 0.1963
Epoch   2 Batch  362/538 - Train Accuracy: 0.8477, Validation Accuracy: 0.8429, Loss: 0.1848
Epoch   2 Batch  363/538 - Train Accuracy: 0.8397, Validation Accuracy

Epoch   2 Batch  442/538 - Train Accuracy: 0.8588, Validation Accuracy: 0.8311, Loss: 0.1692
Epoch   2 Batch  443/538 - Train Accuracy: 0.8728, Validation Accuracy: 0.8256, Loss: 0.1769
Epoch   2 Batch  444/538 - Train Accuracy: 0.8562, Validation Accuracy: 0.8278, Loss: 0.1688
Epoch   2 Batch  445/538 - Train Accuracy: 0.8621, Validation Accuracy: 0.8265, Loss: 0.1711
Epoch   2 Batch  446/538 - Train Accuracy: 0.8690, Validation Accuracy: 0.8249, Loss: 0.1621
Epoch   2 Batch  447/538 - Train Accuracy: 0.8225, Validation Accuracy: 0.8249, Loss: 0.1890
Epoch   2 Batch  448/538 - Train Accuracy: 0.8644, Validation Accuracy: 0.8184, Loss: 0.1638
Epoch   2 Batch  449/538 - Train Accuracy: 0.8537, Validation Accuracy: 0.8235, Loss: 0.1834
Epoch   2 Batch  450/538 - Train Accuracy: 0.8542, Validation Accuracy: 0.8329, Loss: 0.1962
Epoch   2 Batch  451/538 - Train Accuracy: 0.8621, Validation Accuracy: 0.8232, Loss: 0.1656
Epoch   2 Batch  452/538 - Train Accuracy: 0.8689, Validation Accuracy

Epoch   2 Batch  531/538 - Train Accuracy: 0.8613, Validation Accuracy: 0.8411, Loss: 0.1582
Epoch   2 Batch  532/538 - Train Accuracy: 0.8625, Validation Accuracy: 0.8349, Loss: 0.1595
Epoch   2 Batch  533/538 - Train Accuracy: 0.8578, Validation Accuracy: 0.8285, Loss: 0.1600
Epoch   2 Batch  534/538 - Train Accuracy: 0.8642, Validation Accuracy: 0.8347, Loss: 0.1557
Epoch   2 Batch  535/538 - Train Accuracy: 0.8499, Validation Accuracy: 0.8281, Loss: 0.1524
Epoch   2 Batch  536/538 - Train Accuracy: 0.8847, Validation Accuracy: 0.8324, Loss: 0.1711
Epoch   3 Batch    1/538 - Train Accuracy: 0.8396, Validation Accuracy: 0.8333, Loss: 0.1645
Epoch   3 Batch    2/538 - Train Accuracy: 0.8691, Validation Accuracy: 0.8459, Loss: 0.1696
Epoch   3 Batch    3/538 - Train Accuracy: 0.8424, Validation Accuracy: 0.8409, Loss: 0.1614
Epoch   3 Batch    4/538 - Train Accuracy: 0.8498, Validation Accuracy: 0.8317, Loss: 0.1658
Epoch   3 Batch    5/538 - Train Accuracy: 0.8571, Validation Accuracy

Epoch   3 Batch   84/538 - Train Accuracy: 0.8557, Validation Accuracy: 0.8457, Loss: 0.1642
Epoch   3 Batch   85/538 - Train Accuracy: 0.8771, Validation Accuracy: 0.8336, Loss: 0.1375
Epoch   3 Batch   86/538 - Train Accuracy: 0.8646, Validation Accuracy: 0.8452, Loss: 0.1586
Epoch   3 Batch   87/538 - Train Accuracy: 0.8820, Validation Accuracy: 0.8418, Loss: 0.1533
Epoch   3 Batch   88/538 - Train Accuracy: 0.8846, Validation Accuracy: 0.8471, Loss: 0.1495
Epoch   3 Batch   89/538 - Train Accuracy: 0.8707, Validation Accuracy: 0.8441, Loss: 0.1448
Epoch   3 Batch   90/538 - Train Accuracy: 0.8689, Validation Accuracy: 0.8429, Loss: 0.1580
Epoch   3 Batch   91/538 - Train Accuracy: 0.8785, Validation Accuracy: 0.8480, Loss: 0.1519
Epoch   3 Batch   92/538 - Train Accuracy: 0.8410, Validation Accuracy: 0.8514, Loss: 0.1584
Epoch   3 Batch   93/538 - Train Accuracy: 0.8480, Validation Accuracy: 0.8453, Loss: 0.1429
Epoch   3 Batch   94/538 - Train Accuracy: 0.8676, Validation Accuracy

Epoch   3 Batch  173/538 - Train Accuracy: 0.8795, Validation Accuracy: 0.8343, Loss: 0.1280
Epoch   3 Batch  174/538 - Train Accuracy: 0.8521, Validation Accuracy: 0.8420, Loss: 0.1404
Epoch   3 Batch  175/538 - Train Accuracy: 0.8807, Validation Accuracy: 0.8430, Loss: 0.1338
Epoch   3 Batch  176/538 - Train Accuracy: 0.8682, Validation Accuracy: 0.8512, Loss: 0.1495
Epoch   3 Batch  177/538 - Train Accuracy: 0.8566, Validation Accuracy: 0.8549, Loss: 0.1379
Epoch   3 Batch  178/538 - Train Accuracy: 0.8542, Validation Accuracy: 0.8553, Loss: 0.1305
Epoch   3 Batch  179/538 - Train Accuracy: 0.8805, Validation Accuracy: 0.8581, Loss: 0.1374
Epoch   3 Batch  180/538 - Train Accuracy: 0.8979, Validation Accuracy: 0.8647, Loss: 0.1364
Epoch   3 Batch  181/538 - Train Accuracy: 0.8570, Validation Accuracy: 0.8702, Loss: 0.1528
Epoch   3 Batch  182/538 - Train Accuracy: 0.8814, Validation Accuracy: 0.8649, Loss: 0.1257
Epoch   3 Batch  183/538 - Train Accuracy: 0.8873, Validation Accuracy

Epoch   3 Batch  262/538 - Train Accuracy: 0.8902, Validation Accuracy: 0.8535, Loss: 0.1271
Epoch   3 Batch  263/538 - Train Accuracy: 0.8611, Validation Accuracy: 0.8512, Loss: 0.1301
Epoch   3 Batch  264/538 - Train Accuracy: 0.8963, Validation Accuracy: 0.8532, Loss: 0.1325
Epoch   3 Batch  265/538 - Train Accuracy: 0.8736, Validation Accuracy: 0.8608, Loss: 0.1402
Epoch   3 Batch  266/538 - Train Accuracy: 0.8834, Validation Accuracy: 0.8702, Loss: 0.1350
Epoch   3 Batch  267/538 - Train Accuracy: 0.8750, Validation Accuracy: 0.8668, Loss: 0.1339
Epoch   3 Batch  268/538 - Train Accuracy: 0.8949, Validation Accuracy: 0.8835, Loss: 0.1233
Epoch   3 Batch  269/538 - Train Accuracy: 0.8912, Validation Accuracy: 0.8732, Loss: 0.1315
Epoch   3 Batch  270/538 - Train Accuracy: 0.8736, Validation Accuracy: 0.8697, Loss: 0.1317
Epoch   3 Batch  271/538 - Train Accuracy: 0.8727, Validation Accuracy: 0.8704, Loss: 0.1280
Epoch   3 Batch  272/538 - Train Accuracy: 0.8588, Validation Accuracy

Epoch   3 Batch  351/538 - Train Accuracy: 0.8951, Validation Accuracy: 0.8775, Loss: 0.1410
Epoch   3 Batch  352/538 - Train Accuracy: 0.8586, Validation Accuracy: 0.8690, Loss: 0.1438
Epoch   3 Batch  353/538 - Train Accuracy: 0.8859, Validation Accuracy: 0.8720, Loss: 0.1326
Epoch   3 Batch  354/538 - Train Accuracy: 0.8904, Validation Accuracy: 0.8796, Loss: 0.1335
Epoch   3 Batch  355/538 - Train Accuracy: 0.8949, Validation Accuracy: 0.8858, Loss: 0.1356
Epoch   3 Batch  356/538 - Train Accuracy: 0.9114, Validation Accuracy: 0.8839, Loss: 0.1132
Epoch   3 Batch  357/538 - Train Accuracy: 0.9152, Validation Accuracy: 0.8855, Loss: 0.1054
Epoch   3 Batch  358/538 - Train Accuracy: 0.9092, Validation Accuracy: 0.8901, Loss: 0.1177
Epoch   3 Batch  359/538 - Train Accuracy: 0.8869, Validation Accuracy: 0.8841, Loss: 0.1266
Epoch   3 Batch  360/538 - Train Accuracy: 0.9000, Validation Accuracy: 0.8825, Loss: 0.1185
Epoch   3 Batch  361/538 - Train Accuracy: 0.9167, Validation Accuracy

Epoch   3 Batch  440/538 - Train Accuracy: 0.9043, Validation Accuracy: 0.9025, Loss: 0.1217
Epoch   3 Batch  441/538 - Train Accuracy: 0.9133, Validation Accuracy: 0.8984, Loss: 0.1286
Epoch   3 Batch  442/538 - Train Accuracy: 0.9134, Validation Accuracy: 0.8851, Loss: 0.1080
Epoch   3 Batch  443/538 - Train Accuracy: 0.9059, Validation Accuracy: 0.8796, Loss: 0.1202
Epoch   3 Batch  444/538 - Train Accuracy: 0.9107, Validation Accuracy: 0.8917, Loss: 0.1179
Epoch   3 Batch  445/538 - Train Accuracy: 0.9047, Validation Accuracy: 0.8887, Loss: 0.0997
Epoch   3 Batch  446/538 - Train Accuracy: 0.9260, Validation Accuracy: 0.8770, Loss: 0.1073
Epoch   3 Batch  447/538 - Train Accuracy: 0.8842, Validation Accuracy: 0.8754, Loss: 0.1150
Epoch   3 Batch  448/538 - Train Accuracy: 0.8988, Validation Accuracy: 0.8793, Loss: 0.1152
Epoch   3 Batch  449/538 - Train Accuracy: 0.9201, Validation Accuracy: 0.8819, Loss: 0.1253
Epoch   3 Batch  450/538 - Train Accuracy: 0.8891, Validation Accuracy

Epoch   3 Batch  529/538 - Train Accuracy: 0.8928, Validation Accuracy: 0.8979, Loss: 0.1123
Epoch   3 Batch  530/538 - Train Accuracy: 0.9025, Validation Accuracy: 0.8997, Loss: 0.1248
Epoch   3 Batch  531/538 - Train Accuracy: 0.9193, Validation Accuracy: 0.9006, Loss: 0.1092
Epoch   3 Batch  532/538 - Train Accuracy: 0.9270, Validation Accuracy: 0.9002, Loss: 0.0973
Epoch   3 Batch  533/538 - Train Accuracy: 0.9174, Validation Accuracy: 0.8999, Loss: 0.1051
Epoch   3 Batch  534/538 - Train Accuracy: 0.9317, Validation Accuracy: 0.9027, Loss: 0.1008
Epoch   3 Batch  535/538 - Train Accuracy: 0.9135, Validation Accuracy: 0.9045, Loss: 0.0966
Epoch   3 Batch  536/538 - Train Accuracy: 0.9351, Validation Accuracy: 0.9015, Loss: 0.1140
Epoch   4 Batch    1/538 - Train Accuracy: 0.9201, Validation Accuracy: 0.8908, Loss: 0.0974
Epoch   4 Batch    2/538 - Train Accuracy: 0.9180, Validation Accuracy: 0.8910, Loss: 0.1208
Epoch   4 Batch    3/538 - Train Accuracy: 0.9201, Validation Accuracy

Epoch   4 Batch   82/538 - Train Accuracy: 0.9129, Validation Accuracy: 0.9073, Loss: 0.1011
Epoch   4 Batch   83/538 - Train Accuracy: 0.8994, Validation Accuracy: 0.9068, Loss: 0.1042
Epoch   4 Batch   84/538 - Train Accuracy: 0.9161, Validation Accuracy: 0.9137, Loss: 0.1062
Epoch   4 Batch   85/538 - Train Accuracy: 0.9416, Validation Accuracy: 0.9118, Loss: 0.0884
Epoch   4 Batch   86/538 - Train Accuracy: 0.9189, Validation Accuracy: 0.9064, Loss: 0.1036
Epoch   4 Batch   87/538 - Train Accuracy: 0.9143, Validation Accuracy: 0.8929, Loss: 0.0962
Epoch   4 Batch   88/538 - Train Accuracy: 0.9271, Validation Accuracy: 0.9013, Loss: 0.0976
Epoch   4 Batch   89/538 - Train Accuracy: 0.9156, Validation Accuracy: 0.8947, Loss: 0.1040
Epoch   4 Batch   90/538 - Train Accuracy: 0.9182, Validation Accuracy: 0.8999, Loss: 0.1150
Epoch   4 Batch   91/538 - Train Accuracy: 0.9195, Validation Accuracy: 0.9004, Loss: 0.1066
Epoch   4 Batch   92/538 - Train Accuracy: 0.9260, Validation Accuracy

Epoch   4 Batch  171/538 - Train Accuracy: 0.9236, Validation Accuracy: 0.8944, Loss: 0.0939
Epoch   4 Batch  172/538 - Train Accuracy: 0.9159, Validation Accuracy: 0.8924, Loss: 0.0871
Epoch   4 Batch  173/538 - Train Accuracy: 0.9273, Validation Accuracy: 0.8940, Loss: 0.0809
Epoch   4 Batch  174/538 - Train Accuracy: 0.9141, Validation Accuracy: 0.9068, Loss: 0.0908
Epoch   4 Batch  175/538 - Train Accuracy: 0.9287, Validation Accuracy: 0.9107, Loss: 0.0878
Epoch   4 Batch  176/538 - Train Accuracy: 0.9023, Validation Accuracy: 0.9047, Loss: 0.1047
Epoch   4 Batch  177/538 - Train Accuracy: 0.9183, Validation Accuracy: 0.9096, Loss: 0.0986
Epoch   4 Batch  178/538 - Train Accuracy: 0.8858, Validation Accuracy: 0.9103, Loss: 0.0950
Epoch   4 Batch  179/538 - Train Accuracy: 0.9137, Validation Accuracy: 0.9091, Loss: 0.0960
Epoch   4 Batch  180/538 - Train Accuracy: 0.9267, Validation Accuracy: 0.9102, Loss: 0.0946
Epoch   4 Batch  181/538 - Train Accuracy: 0.9215, Validation Accuracy

Epoch   4 Batch  260/538 - Train Accuracy: 0.9027, Validation Accuracy: 0.9148, Loss: 0.0913
Epoch   4 Batch  261/538 - Train Accuracy: 0.9240, Validation Accuracy: 0.9215, Loss: 0.1005
Epoch   4 Batch  262/538 - Train Accuracy: 0.9359, Validation Accuracy: 0.9206, Loss: 0.0814
Epoch   4 Batch  263/538 - Train Accuracy: 0.9141, Validation Accuracy: 0.9187, Loss: 0.0870
Epoch   4 Batch  264/538 - Train Accuracy: 0.9262, Validation Accuracy: 0.9160, Loss: 0.0918
Epoch   4 Batch  265/538 - Train Accuracy: 0.9211, Validation Accuracy: 0.9187, Loss: 0.0989
Epoch   4 Batch  266/538 - Train Accuracy: 0.9120, Validation Accuracy: 0.9251, Loss: 0.0887
Epoch   4 Batch  267/538 - Train Accuracy: 0.9283, Validation Accuracy: 0.9224, Loss: 0.0887
Epoch   4 Batch  268/538 - Train Accuracy: 0.9273, Validation Accuracy: 0.9279, Loss: 0.0790
Epoch   4 Batch  269/538 - Train Accuracy: 0.9297, Validation Accuracy: 0.9206, Loss: 0.0969
Epoch   4 Batch  270/538 - Train Accuracy: 0.9174, Validation Accuracy

Epoch   4 Batch  349/538 - Train Accuracy: 0.9502, Validation Accuracy: 0.9078, Loss: 0.0823
Epoch   4 Batch  350/538 - Train Accuracy: 0.9306, Validation Accuracy: 0.9123, Loss: 0.0893
Epoch   4 Batch  351/538 - Train Accuracy: 0.9141, Validation Accuracy: 0.9137, Loss: 0.0952
Epoch   4 Batch  352/538 - Train Accuracy: 0.9076, Validation Accuracy: 0.9180, Loss: 0.1087
Epoch   4 Batch  353/538 - Train Accuracy: 0.9223, Validation Accuracy: 0.9252, Loss: 0.0855
Epoch   4 Batch  354/538 - Train Accuracy: 0.9145, Validation Accuracy: 0.9224, Loss: 0.0859
Epoch   4 Batch  355/538 - Train Accuracy: 0.9287, Validation Accuracy: 0.9229, Loss: 0.0866
Epoch   4 Batch  356/538 - Train Accuracy: 0.9343, Validation Accuracy: 0.9217, Loss: 0.0738
Epoch   4 Batch  357/538 - Train Accuracy: 0.9432, Validation Accuracy: 0.9229, Loss: 0.0820
Epoch   4 Batch  358/538 - Train Accuracy: 0.9313, Validation Accuracy: 0.9228, Loss: 0.0692
Epoch   4 Batch  359/538 - Train Accuracy: 0.9254, Validation Accuracy

Epoch   4 Batch  438/538 - Train Accuracy: 0.9295, Validation Accuracy: 0.9261, Loss: 0.0696
Epoch   4 Batch  439/538 - Train Accuracy: 0.9461, Validation Accuracy: 0.9229, Loss: 0.0792
Epoch   4 Batch  440/538 - Train Accuracy: 0.9229, Validation Accuracy: 0.9217, Loss: 0.0846
Epoch   4 Batch  441/538 - Train Accuracy: 0.9158, Validation Accuracy: 0.9196, Loss: 0.0960
Epoch   4 Batch  442/538 - Train Accuracy: 0.9409, Validation Accuracy: 0.9180, Loss: 0.0693
Epoch   4 Batch  443/538 - Train Accuracy: 0.9228, Validation Accuracy: 0.9213, Loss: 0.0800
Epoch   4 Batch  444/538 - Train Accuracy: 0.9351, Validation Accuracy: 0.9238, Loss: 0.0839
Epoch   4 Batch  445/538 - Train Accuracy: 0.9377, Validation Accuracy: 0.9277, Loss: 0.0765
Epoch   4 Batch  446/538 - Train Accuracy: 0.9373, Validation Accuracy: 0.9237, Loss: 0.0798
Epoch   4 Batch  447/538 - Train Accuracy: 0.9148, Validation Accuracy: 0.9114, Loss: 0.0769
Epoch   4 Batch  448/538 - Train Accuracy: 0.9401, Validation Accuracy

Epoch   4 Batch  527/538 - Train Accuracy: 0.9361, Validation Accuracy: 0.9284, Loss: 0.0757
Epoch   4 Batch  528/538 - Train Accuracy: 0.9285, Validation Accuracy: 0.9281, Loss: 0.0798
Epoch   4 Batch  529/538 - Train Accuracy: 0.9184, Validation Accuracy: 0.9277, Loss: 0.0749
Epoch   4 Batch  530/538 - Train Accuracy: 0.9072, Validation Accuracy: 0.9313, Loss: 0.0843
Epoch   4 Batch  531/538 - Train Accuracy: 0.9322, Validation Accuracy: 0.9256, Loss: 0.0746
Epoch   4 Batch  532/538 - Train Accuracy: 0.9512, Validation Accuracy: 0.9292, Loss: 0.0647
Epoch   4 Batch  533/538 - Train Accuracy: 0.9400, Validation Accuracy: 0.9327, Loss: 0.0735
Epoch   4 Batch  534/538 - Train Accuracy: 0.9343, Validation Accuracy: 0.9308, Loss: 0.0554
Epoch   4 Batch  535/538 - Train Accuracy: 0.9330, Validation Accuracy: 0.9316, Loss: 0.0678
Epoch   4 Batch  536/538 - Train Accuracy: 0.9440, Validation Accuracy: 0.9316, Loss: 0.0848
Epoch   5 Batch    1/538 - Train Accuracy: 0.9510, Validation Accuracy

Epoch   5 Batch   80/538 - Train Accuracy: 0.9477, Validation Accuracy: 0.9446, Loss: 0.0784
Epoch   5 Batch   81/538 - Train Accuracy: 0.9428, Validation Accuracy: 0.9448, Loss: 0.0730
Epoch   5 Batch   82/538 - Train Accuracy: 0.9250, Validation Accuracy: 0.9377, Loss: 0.0699
Epoch   5 Batch   83/538 - Train Accuracy: 0.9229, Validation Accuracy: 0.9403, Loss: 0.0652
Epoch   5 Batch   84/538 - Train Accuracy: 0.9191, Validation Accuracy: 0.9379, Loss: 0.0668
Epoch   5 Batch   85/538 - Train Accuracy: 0.9590, Validation Accuracy: 0.9350, Loss: 0.0594
Epoch   5 Batch   86/538 - Train Accuracy: 0.9520, Validation Accuracy: 0.9336, Loss: 0.0653
Epoch   5 Batch   87/538 - Train Accuracy: 0.9412, Validation Accuracy: 0.9343, Loss: 0.0737
Epoch   5 Batch   88/538 - Train Accuracy: 0.9545, Validation Accuracy: 0.9338, Loss: 0.0632
Epoch   5 Batch   89/538 - Train Accuracy: 0.9303, Validation Accuracy: 0.9348, Loss: 0.0623
Epoch   5 Batch   90/538 - Train Accuracy: 0.9442, Validation Accuracy

Epoch   5 Batch  169/538 - Train Accuracy: 0.9596, Validation Accuracy: 0.9352, Loss: 0.0537
Epoch   5 Batch  170/538 - Train Accuracy: 0.9394, Validation Accuracy: 0.9316, Loss: 0.0697
Epoch   5 Batch  171/538 - Train Accuracy: 0.9219, Validation Accuracy: 0.9292, Loss: 0.0693
Epoch   5 Batch  172/538 - Train Accuracy: 0.9328, Validation Accuracy: 0.9228, Loss: 0.0568
Epoch   5 Batch  173/538 - Train Accuracy: 0.9453, Validation Accuracy: 0.9176, Loss: 0.0539
Epoch   5 Batch  174/538 - Train Accuracy: 0.9246, Validation Accuracy: 0.9201, Loss: 0.0641
Epoch   5 Batch  175/538 - Train Accuracy: 0.9518, Validation Accuracy: 0.9256, Loss: 0.0634
Epoch   5 Batch  176/538 - Train Accuracy: 0.9260, Validation Accuracy: 0.9297, Loss: 0.0707
Epoch   5 Batch  177/538 - Train Accuracy: 0.9418, Validation Accuracy: 0.9411, Loss: 0.0601
Epoch   5 Batch  178/538 - Train Accuracy: 0.9139, Validation Accuracy: 0.9396, Loss: 0.0740
Epoch   5 Batch  179/538 - Train Accuracy: 0.9418, Validation Accuracy

Epoch   5 Batch  258/538 - Train Accuracy: 0.9537, Validation Accuracy: 0.9364, Loss: 0.0615
Epoch   5 Batch  259/538 - Train Accuracy: 0.9533, Validation Accuracy: 0.9332, Loss: 0.0547
Epoch   5 Batch  260/538 - Train Accuracy: 0.9122, Validation Accuracy: 0.9387, Loss: 0.0676
Epoch   5 Batch  261/538 - Train Accuracy: 0.9342, Validation Accuracy: 0.9387, Loss: 0.0712
Epoch   5 Batch  262/538 - Train Accuracy: 0.9537, Validation Accuracy: 0.9471, Loss: 0.0597
Epoch   5 Batch  263/538 - Train Accuracy: 0.9389, Validation Accuracy: 0.9375, Loss: 0.0590
Epoch   5 Batch  264/538 - Train Accuracy: 0.9314, Validation Accuracy: 0.9451, Loss: 0.0662
Epoch   5 Batch  265/538 - Train Accuracy: 0.9270, Validation Accuracy: 0.9419, Loss: 0.0733
Epoch   5 Batch  266/538 - Train Accuracy: 0.9144, Validation Accuracy: 0.9419, Loss: 0.0733
Epoch   5 Batch  267/538 - Train Accuracy: 0.9432, Validation Accuracy: 0.9389, Loss: 0.0582
Epoch   5 Batch  268/538 - Train Accuracy: 0.9368, Validation Accuracy

Epoch   5 Batch  347/538 - Train Accuracy: 0.9395, Validation Accuracy: 0.9396, Loss: 0.0618
Epoch   5 Batch  348/538 - Train Accuracy: 0.9414, Validation Accuracy: 0.9407, Loss: 0.0531
Epoch   5 Batch  349/538 - Train Accuracy: 0.9619, Validation Accuracy: 0.9409, Loss: 0.0510
Epoch   5 Batch  350/538 - Train Accuracy: 0.9505, Validation Accuracy: 0.9419, Loss: 0.0671
Epoch   5 Batch  351/538 - Train Accuracy: 0.9459, Validation Accuracy: 0.9450, Loss: 0.0744
Epoch   5 Batch  352/538 - Train Accuracy: 0.9208, Validation Accuracy: 0.9492, Loss: 0.0791
Epoch   5 Batch  353/538 - Train Accuracy: 0.9385, Validation Accuracy: 0.9485, Loss: 0.0645
Epoch   5 Batch  354/538 - Train Accuracy: 0.9271, Validation Accuracy: 0.9483, Loss: 0.0573
Epoch   5 Batch  355/538 - Train Accuracy: 0.9535, Validation Accuracy: 0.9494, Loss: 0.0666
Epoch   5 Batch  356/538 - Train Accuracy: 0.9494, Validation Accuracy: 0.9521, Loss: 0.0540
Epoch   5 Batch  357/538 - Train Accuracy: 0.9553, Validation Accuracy

Epoch   5 Batch  436/538 - Train Accuracy: 0.9314, Validation Accuracy: 0.9451, Loss: 0.0695
Epoch   5 Batch  437/538 - Train Accuracy: 0.9492, Validation Accuracy: 0.9437, Loss: 0.0596
Epoch   5 Batch  438/538 - Train Accuracy: 0.9570, Validation Accuracy: 0.9393, Loss: 0.0543
Epoch   5 Batch  439/538 - Train Accuracy: 0.9545, Validation Accuracy: 0.9308, Loss: 0.0591
Epoch   5 Batch  440/538 - Train Accuracy: 0.9344, Validation Accuracy: 0.9309, Loss: 0.0662
Epoch   5 Batch  441/538 - Train Accuracy: 0.9324, Validation Accuracy: 0.9411, Loss: 0.0736
Epoch   5 Batch  442/538 - Train Accuracy: 0.9675, Validation Accuracy: 0.9382, Loss: 0.0491
Epoch   5 Batch  443/538 - Train Accuracy: 0.9395, Validation Accuracy: 0.9366, Loss: 0.0609
Epoch   5 Batch  444/538 - Train Accuracy: 0.9522, Validation Accuracy: 0.9263, Loss: 0.0517
Epoch   5 Batch  445/538 - Train Accuracy: 0.9518, Validation Accuracy: 0.9245, Loss: 0.0526
Epoch   5 Batch  446/538 - Train Accuracy: 0.9546, Validation Accuracy

Epoch   5 Batch  525/538 - Train Accuracy: 0.9462, Validation Accuracy: 0.9329, Loss: 0.0655
Epoch   5 Batch  526/538 - Train Accuracy: 0.9542, Validation Accuracy: 0.9370, Loss: 0.0597
Epoch   5 Batch  527/538 - Train Accuracy: 0.9381, Validation Accuracy: 0.9405, Loss: 0.0547
Epoch   5 Batch  528/538 - Train Accuracy: 0.9359, Validation Accuracy: 0.9398, Loss: 0.0677
Epoch   5 Batch  529/538 - Train Accuracy: 0.9170, Validation Accuracy: 0.9423, Loss: 0.0655
Epoch   5 Batch  530/538 - Train Accuracy: 0.9119, Validation Accuracy: 0.9366, Loss: 0.0639
Epoch   5 Batch  531/538 - Train Accuracy: 0.9311, Validation Accuracy: 0.9384, Loss: 0.0603
Epoch   5 Batch  532/538 - Train Accuracy: 0.9426, Validation Accuracy: 0.9448, Loss: 0.0505
Epoch   5 Batch  533/538 - Train Accuracy: 0.9539, Validation Accuracy: 0.9448, Loss: 0.0604
Epoch   5 Batch  534/538 - Train Accuracy: 0.9462, Validation Accuracy: 0.9442, Loss: 0.0447
Epoch   5 Batch  535/538 - Train Accuracy: 0.9576, Validation Accuracy

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [47]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [48]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [49]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    # TODO: Implement Function
    word_ids = []
    for word in sentence.lower().split():
        if word in vocab_to_int:
            word_ids.append(vocab_to_int[word])
        else:
            word_ids.append(vocab_to_int['<UNK>'])
    return word_ids


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [53]:
translate_sentence = 'he saw a old yellow truck .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [102, 12, 131, 48, 204, 22, 83]
  English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']

Prediction
  Word Ids:      [205, 62, 220, 132, 12, 278, 114, 38, 1]
  French Words: il a vu un gros camion jaune . <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.